osquery/events/linux/containereventpublisher.cpp

```diff
@@ -0,0 +1,560 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <regex>
+#include <unordered_map>
+
+#include <boost/asio.hpp>
+
+#include <grpcpp/grpcpp.h>
+
+#include <google/protobuf/util/time_util.h>
+
+#include <osquery/core/flags.h>
+#include <osquery/events/containerd/container.grpc.pb.h>
+#include <osquery/events/containerd/content.grpc.pb.h>
+#include <osquery/events/containerd/image.grpc.pb.h>
+#include <osquery/events/containerd/namespace.grpc.pb.h>
+#include <osquery/events/containerd/runc.grpc.pb.h>
+#include <osquery/events/containerd/snapshot.grpc.pb.h>
+#include <osquery/events/containerd/task.grpc.pb.h>
+#include <osquery/events/eventsubscriber.h>
+#include <osquery/events/linux/containereventpublisher.h>
+#include <osquery/logger/logger.h>
+#include <osquery/registry/registry_factory.h>
+#include <osquery/utils/json/json.h>
+
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+namespace local = boost::asio::local;
+
+namespace osquery {
+
+const std::string kContainerEventTypePrefix = "containerd.events.Container";
+const std::string kContentEventTypePrefix = "containerd.events.Content";
+const std::string kImageEventTypePrefix = "containerd.events.Image";
+const std::string kNamespaceEventTypePrefix = "containerd.events.Namespace";
+const std::string kSnapshotEventTypePrefix = "containerd.events.Snapshot";
+const std::string kTaskEventTypePrefix = "containerd.events.Task";
+
+REGISTER(ContainerEventPublisher, "event_publisher", "containerevent");
+
+FLAG(string,
+     containerd_socket,
+     "/run/containerd/containerd.sock",
+     "Docker UNIX domain socket path");
+
+FLAG(string,
+     containerd_events_namespaces,
+     "",
+     "Comma separated list of namespaces to receive events from, empty means "
+     "all");
+
+FLAG(bool,
+     enable_containerd_events_publisher,
+     false,
+     "Enable the containerd event publisher");
```

Comment:
let's make sure we document these in the docs https://github.com/osquery/osquery/tree/master/docs

Reply:
Documented in cli-flags.md. I suspect a larger documentation section on containerd events would be useful but I might hold of on that until we have more tables (e.g. also non-evented ones)